website/events-listener/src/events.ts

```diff
@@ -21,100 +22,122 @@ import { logger } from './logger';
+      callback({
+        key: Keys.dispatchMessage,
+        value: {
+          ...base,
+          programId: eventData.programId.toHex(),
+          messageId: eventData.messageId.toHex(),
+          origin: eventData.origin.toHex(),
+        },
+      } as AddEventKafkaPayload<Keys.dispatchMessage, DispatchMessageEnqueud>);
+      break;
+    case 'Log':
+      eventData = new LogData(data);
+      callback({
+        key: Keys.log,
+        value: {
+          ...base,
+          id: eventData.id.toHex(),
+          source: eventData.source.toHex(),
+          destination: eventData.dest.toHex(),
+          payload: eventData.payload.toHex(),
+          replyTo: eventData.reply.isSome ? eventData.reply.unwrap()[0].toHex() : null,
+          replyError: eventData.reply.isSome ? `${eventData.reply.unwrap()[1].toNumber()}` : null,
+        },
+      } as AddEventKafkaPayload<Keys.log, Log>);
+      break;
+    case 'InitSuccess':
+      eventData = new InitSuccessData(data);
+      callback({
+        key: Keys.initSuccess,
+        value: {
+          ...base,
+          programId: eventData.programId.toHex(),
+          messageId: eventData.messageId.toHex(),
+          origin: eventData.origin.toHex(),
+        },
+      } as AddEventKafkaPayload<Keys.initSuccess, InitSuccess>);
+      break;
+    case 'InitFailure':
+      eventData = new InitFailureData(data);
+      callback({
+        key: Keys.initFailure,
+        value: {
+          ...base,
+          programId: eventData.info.programId.toHex(),
+          messageId: eventData.info.messageId.toHex(),
+          origin: eventData.info.origin.toHex(),
+        },
+      } as AddEventKafkaPayload<Keys.initFailure, InitFailure>);
+      break;
+    case 'MessageDispatched':
+      eventData = new MessageDispatchedData(data);
+      callback({
+        key: Keys.messageDispatched,
+        value: {
+          ...base,
+          messageId: eventData.messageId.toHex(),
+          outcome: eventData.outcome.isFailure ? eventData.outcome.asFailure.toHuman() : 'success',
+        },
+      } as AddEventKafkaPayload<Keys.messageDispatched, MessageDispatched>);
+      break;
+    default:
+      console.warn(`unexpected event type to handle: ${method}`);
+  }
+};
+
+export const listen = async (api: GearApi, genesis: string, callback: (arg: { key: string; value: any }) => void) => {
   api.allEvents((events: any) => {
+    const blockHash = events.createdAtHash.toHex();
     const base = {
       genesis,
-      blockHash: events.createdAtHash.toHex(),
-      timestamp: Date.now(),
+      blockHash,
+      timestamp: 0,
     };
 
     events.forEach(async ({ event: { data, method } }: any) => {
-      let eventData:
-        | InitMessageEnqueuedData
-        | DispatchMessageEnqueuedData
-        | LogData
-        | InitSuccessData
-        | InitFailureData
-        | MessageDispatchedData;
+      base.timestamp = (await api.blocks.getBlockTimestamp(blockHash)).toNumber();
```

Comment:
Why didn't you add this value where `base` var was declared?

Reply:
A silly oversight of me, now fixed. Thanks.